package/run.sh

```diff
@@ -31,16 +31,14 @@ do
         echo "not a K3s server, not attempting error remediation"
         exit 1
     fi
-    # We need to get the right systemd service based off of the role we have, $INSTALL_K3S_EXEC will be set if we are a worker only node, in which case the systemd service is k3s-agent.service
-    journalctl -u k3s.service > k3s-service.txt
-    # Only use the logs for the latest restart of k3s, otherwise errors encountered after an ETCD join error will not be reported properly.
-    LAST_START_LINE=$(cat k3s-service.txt | grep "Starting k3s v1." -n  | cut -d: -f1 | tail -1)
-    if [[ "$LAST_START_LINE" -gt 0 ]]; then
-        LAST_LOGS=$(cat k3s-service.txt | sed -n "$((LAST_START_LINE))"',$p')
-        else
-        LAST_LOGS=$(cat k3s-service.txt)
-    fi
-    if ! echo "${LAST_LOGS}" | grep "ETCD join failed: etcdserver: too many learner members in cluster" -q && echo $?; then
+
+    # Get the last time the service started using the ExecMainStartTimestamp property
+    START=$(systemctl show k3s.service --property=ExecMainStartTimestamp | cut -f2 -d"=")
+    # Convert the last start timestamp into rfc-3339 so we can feed it to journalctl
+    DATE=$(date -d "$START" --rfc-3339=seconds | cut -f1 -d"+")
```

Comment:
1. The `date` command isn't necessary - the format returned by `systemctl show` timestamp fields is parsable by `journactl --since` - see the docs and confirmed by Lennart himself [in a comment](https://github.com/systemd/systemd/issues/22567#issuecomment-1045257778). 
    ```
    -S, --since=, -U, --until=
    Start showing entries on or newer than the specified date, or on or older than the specified date, respectively.
    Date specifications should be of the format "2012-10-30 18:17:16".
    ```
2. Lets just search for the [bare etcd error](https://pkg.go.dev/go.etcd.io/etcd/api/v3/v3rpc/rpctypes#ErrGRPCTooManyLearners), as the k3s framing may change over time.
3. Negating the test makes the code less readable. If the specific error is found then sleep, if not then exit.

Putting it all together, perhaps something like this?

```bash
    # Get the last time the service started using the ExecMainStartTimestamp property
    START=$(systemctl show k3s.service --property=ExecMainStartTimestamp | cut -f2 -d=)
    # Get the logs since the last time the service started and check for the given error
    if journalctl -u k3s.service --since="${START}" | grep -qF "etcdserver: too many learner members in cluster"; then
        # We couldn't register as a learner, keep trying until we can
        sleep "${DELAY}"
    else
        # if we have an error that isn't an etcd learner member error we shouldn't retry
        exit 1
    fi
```

Reply:
If you're seeing that the timestamp format isn't parsable when testing locally, you may need to disable localization? Try `LANG=C systemctl show k3s.service --property=ExecMainStartTimestamp | cut -f2 -d=`

Reply:
Oh awesome, I think during my initial testing passing the date to `journalctl` I may have missed a `'`, so the parsing failed. I didn't like the date formatting but thought it was necessary, thanks for pointing that out. 

I'll test that code block to make sure it works, but I'm fine with adopting that suggestion wholesale, thanks for providing it.